crates/pipeline_manager/src/runner.rs

```diff
@@ -551,7 +552,500 @@ impl LocalRunner {
+
+    /// How often to poll for the pipeline process to exit.
+    const SHUTDOWN_POLL_PERIOD: Duration = Duration::from_millis(300);
+
+    fn new(
+        pipeline_id: PipelineId,
+        tenant_id: TenantId,
+        config: &Arc<CompilerConfig>,
+        db: Arc<Mutex<ProjectDB>>,
+        notifier: Arc<Notify>,
+    ) -> Self {
+        Self {
+            pipeline_id,
+            tenant_id,
+            pipeline_process: None,
+            config: config.clone(),
+            db,
+            notifier,
+        }
+    }
+
+    /// Runs until the pipeline is deleted or an unexpected error occurs.
+    async fn run(self) -> Result<(), ManagerError> {
+        let pipeline_id = self.pipeline_id;
+
+        self.do_run().await.map_err(|e| {
+            error!(
+                "Pipeline automaton '{}' terminated with error: '{e}'",
+                pipeline_id
+            );
+            e
+        })
+    }
+
+    async fn do_run(mut self) -> Result<(), ManagerError> {
+        let mut poll_timeout = Self::DEFAULT_PIPELINE_POLL_PERIOD;
+
+        loop {
+            // Wait until the timeout expires or we get notified that
+            // the desired state of the pipelime has changed.
+            let _ = timeout(poll_timeout, self.notifier.notified()).await;
+            poll_timeout = Self::DEFAULT_PIPELINE_POLL_PERIOD;
+
+            // TODO: use transactional API when ready to avoid races with a parallel
+            // `/deploy /shutdown` sequence.
+
+            // txn = db.transaction();
+            let db = self.db.lock().await;
+            let mut pipeline = db
+                .get_pipeline_runtime_state(self.tenant_id, self.pipeline_id)
+                .await?;
+
+            // Handle deployment request.
+            if pipeline.current_status == PipelineStatus::Shutdown
+                && pipeline.desired_status != PipelineStatus::Shutdown
+            {
+                pipeline.set_current_status(PipelineStatus::Provisioning, None);
+                let revision = db
+                    .get_last_committed_pipeline_revision(self.tenant_id, self.pipeline_id)
+                    .await?;
+                db.update_pipeline_runtime_state(self.tenant_id, self.pipeline_id, &pipeline)
+                    .await?;
+                // txn.commit();
+                drop(db);
+
+                match self.start(revision).await {
+                    Ok(child) => {
+                        self.pipeline_process = Some(child);
+                    }
+                    Err(e) => {
+                        self.force_kill_pipeline(&mut pipeline, Some(e)).await?;
+                    }
+                }
+                poll_timeout = Self::PROVISIONING_POLL_PERIOD;
+                continue;
+            } else {
+                // txn.commit();
+                drop(db);
+            }
+
+            match (pipeline.current_status, pipeline.desired_status) {
```

Comment:
Each of the arms here could be a method that names the step. E.g., "try_initialize_pipeline()"